### PR TITLE
Introduce models for snapshot/chord for testing

### PR DESCRIPTION
**Pull request by charleso**
Hopefully this is not a step backwards. I would definitely argue the `SnapshotsSpec` has at least improved. `ChordArbitraries` is still bad, but I think we need to "model" squash too before that's really possible.

### CODE REVIEW

File: ivory-core/src/test/scala/com/ambiata/ivory/core/arbitraries/FactsWithStatePriority.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import com.ambiata.ivory.core._
+import com.ambiata.ivory.core.arbitraries.Arbitraries._
+import org.scalacheck.Arbitrary, Arbitrary._
+
+case class FactsWithStatePriority(values: Map[DateTime, NaturalIntSmall]) {
+
+  def factsets(template: Fact): List[List[Fact]] = {
+    val max = values.values.map(_.value).max
+    (0 until max).map(i => values.filter(_._2.value > i).map(x => x._1 -> i).toList).toList
+      // TODO Make value unique
```
Comment (markhibberd) [inline ivory-core/src/test/scala/com/ambiata/ivory/core/arbitraries/FactsWithStatePriority.scala]: Now or later?

File: ivory-operation/src/test/scala/com/ambiata/ivory/operation/model/ChordModel.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  def secondary: Order[S] =
+    SnapshotModel.secondary
+
+  def reduce(key: K, facts: NonEmptyList[Fact], conf: ChordModelConf): List[Fact] =
+    reduceWithDates(facts, conf.dictionary.byConcrete.sources(key._2), conf.entities.entities.get(key._1).toList.map(Date.unsafeFromInt))
+
+  def reduceWithDates(facts: NonEmptyList[Fact], dictionary: ConcreteGroup, dates: List[Date]): List[Fact] = {
+    (Date.minValue :: dates.reverse).sliding(2).toList.flatMap {
+      case List(d1, d2) =>
```
Comment (markhibberd) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/model/ChordModel.scala]: You can't filter a Nel, what happens if you filter out all the elements?

Comment (charleso) [inline ivory-operation/src/test/scala/com/ambiata/ivory/operation/model/ChordModel.scala]: > You can't filter a Nel, what happens if you filter out all the elements?

Doh. Sorry, wasn't thinking.

**Comment by markhibberd** [comment]
:japanese_castle: Looks good, can't see any issues/missing tests.

**Comment by charleso** [comment]
Have added `Value.unique`, will merge when the tests are green.

**Pull request merged**